harness/determined/event_trail/event_trail.py

```diff
@@ -0,0 +1,111 @@
+import threading
+import time
+from typing import Any, Callable, Dict, Generator, Union, List, Type
+from determined.event_trail.event_trail_events import EventTrailEvent, EventPriority, TrialInfoEventV1
+from enum import Enum
+
+
+class bcolors(Enum):
+    HEADER = '\033[95m'
+    OKBLUE = '\033[94m'
+    OKCYAN = '\033[96m'
+    OKGREEN = '\033[92m'
+    WARNING = '\033[93m'
+    FAIL = '\033[91m'
+    ENDC = '\033[0m'
+    BOLD = '\033[1m'
+    UNDERLINE = '\033[4m'
+
+
+def colorize(s: str, color: bcolors):
+    return f"{str(color)}{s}{str(bcolors.ENDC)}"
+
+
+def send_event(event: EventTrailEvent):
+    log_line = colorize(f'EventSenderThread sending: {str(event)}', bcolors.OKGREEN)
+    print(log_line)
+
+
+class EventTrailThread(threading.Thread):
+    """
+    Background thread for sending events to the master asynchronously
+    """
+
+    def __init__(self) -> None:
+        self.events = []  # type: List[EventTrailEvent]
+        self.quitting = False
+        self.main_event_loop_exited = False
+        super().__init__()
+        print("EventSenderThread init")
+
+    def run(self) -> None:
+        while True:
+            print("EventSenderThread While Loop")
+            if self.quitting:
+                break
+
+            if len(self.events) > 0:
+                print("EventSenderThread While Loop Events Exist")
+                self._send_events()
+
+            if len(self.events) == 0:
+                time.sleep(1)
+
+        self.main_event_loop_exited = True
+
+    def enqueue_for_async_send(self, event: EventTrailEvent):
+        self.events.append(event)
```

Comment:
This is not thread-safe.

Check out `queue.Queue` or `queue.PriorityQueue`, they will do most of the heavy lifting for you.